backends/p4test/p4test.cpp

```diff
@@ -100,71 +104,70 @@ int main(int argc, char *const argv[]) {
             error("Can't open %s", options.file); }
     } else {
         program = P4::parseP4File(options);
+
+        if (program != nullptr && ::errorCount() == 0) {
+            P4::P4COptionPragmaParser optionsPragmaParser;
+            program->apply(P4::ApplyOptionsPragmas(optionsPragmaParser));
+
+            if (!options.parseOnly) {
+                try {
+                    P4::FrontEnd fe;
+                    fe.addDebugHook(hook);
+                    program = fe.run(options, program);
+                } catch (const Util::P4CExceptionBase &bug) {
+                    std::cerr << bug.what() << std::endl;
+                    return 1;
+                }
+            }
+        }
     }
-    auto hook = options.getDebugHook();
 
+    log_dump(program, "Initial program");
     if (program != nullptr && ::errorCount() == 0) {
-        P4::P4COptionPragmaParser optionsPragmaParser;
-        program->apply(P4::ApplyOptionsPragmas(optionsPragmaParser));
+        P4::serializeP4RuntimeIfRequired(program, options);
 
-        if (!options.parseOnly) {
+        if (!options.parseOnly && !options.validateOnly) {
+            P4Test::MidEnd midEnd(options);
+            midEnd.addDebugHook(hook);
+#if 0
+            /* doing this breaks the output until we get dump/undump of srcInfo */
+            if (options.debugJson) {
+                std::stringstream tmp;
+                JSONGenerator gen(tmp);
+                gen << program;
+                JSONLoader loader(tmp);
+                loader >> program;
+            }
+#endif
+            const IR::ToplevelBlock *top = nullptr;
             try {
-                P4::FrontEnd fe;
-                fe.addDebugHook(hook);
-                program = fe.run(options, program);
+                top = midEnd.process(program);
             } catch (const Util::P4CExceptionBase &bug) {
                 std::cerr << bug.what() << std::endl;
                 return 1;
             }
+            log_dump(program, "After midend");
+            log_dump(top, "Top level block");
         }
-        log_dump(program, "Initial program");
-        if (program != nullptr && ::errorCount() == 0) {
-            P4::serializeP4RuntimeIfRequired(program, options);
+        if (options.dumpJsonFile)
+            JSONGenerator(*openFile(options.dumpJsonFile, true), true) << program << std::endl;
+        if (options.debugJson) {
```

Comment:
I wonder whether we still need the debugJson option.

Reply:
We use it by default on all p4test tests, so we have regression tests on the toJson/fromJson code.